docs/contributing/effects-tutorial.md

```diff
@@ -0,0 +1,634 @@
+# Effects and How They Work
+
+This doc is intended to provide an approachable reference and tutorial for
+understanding algebraic effect systems in Haskell.  This is not a simple topic,
+but there are a great number of simple facts that can demystify the effect
+system greatly.
+
+This doc is specifically based on the `fused-effects` library, for a number of
+reasons, but mostly because that's the effect system we use in `fossa-cli`.
+
+## Prerequisite Knowledge
+
+In order to understand algebraic effect systems in detail, there is quite a lot
+of things that need to be understood in detail.  If you are new to haskell, or
+just not very confident, please make sure you are familiar with all of these
+concepts.
+
+- Haskell as a language ([learn about this][Haskell101])
+- `Functor`, `Applicative`, and `Monad` ([learn about this][Haskell102])
+- Monad transformers ([learn about this][MTLTutorial])
+
+Once you've become quite familiar with all of that, the last piece of info
+you'll need comes from a talk which explains the motivation for `fused-effects`
+specifically, as opposed to other effect systems.  The talk goes into detail
+on several key points, so make sure you understand the following parts of the
+talk by the end:
+
+- How `mtl`-style monad transformers might work in production code.
+- The benefits of monad transformers, and the guiding principles behind its
+  development.
+- Where `mtl`-style monad transformers break down in some cases.
+  - The _functional dependencies_ point in particular is important.
+- The goals/requirements of the `fused-effects` library.
+
+[Here is the link to the fused-effects talk.][fused-talk]  Pay close attention
+to the points mentioned above or you will almost certainly be confused by why
+`fused-effects` is as complex as it is.
+
+Please also note that the talk is based on a slightly out-of-date version of
+`fused-effects`, and makes references to a `Carrier sig m` typeclass, which
+has been renamed to `Algebra sig m`.
+
+## `fused-effects` At a Glance
+
+Taking a similar example from the `fused-effects` talk, let's come up with a
+simple application that we can write with regular monads, monad transformers,
+and `fused-effects`-style monads.
+
+### Example Program
+
+We'll write a program that read a specific value from a file on the local
+filesystem.  The function to read the value from the file contents comes from a
+third-party library, and we have no idea how it works, other than that it does
+not always contain a valid value, and can therefore return a `Maybe`.
+
+Every time we read a valid value from the file, we'll store that value into a
+ordered buffer, and attempt to read the value again. Once we read an invalid
+value (i.e. the magic file reader returns `Nothing`), we'll stop the loop, and
+print out all of the values we read to stdout.
+
+Let's add one more requirement: we only want to report unique valid values, in
```

Comment:
As I read this through this requirement seems out of place?  Why include this?  This section could use a little explainer of why this example is interesting.  E.g. mixing IO, state and failure.